### PR TITLE
add support for STM32U585

### PR DESCRIPTION
**Pull request by TychoVrahe**
Add support for STM32U585 variant: separate linker scripts and MPU needs different configuration.

### CODE REVIEW

File: core/embed/trezorhal/stm32u5/mpu.c [hunk 1]
```diff
@@ -108,30 +109,51 @@ static void mpu_set_attributes() {
...
 void mpu_config_boardloader() {
   HAL_MPU_Disable();
   mpu_set_attributes();
   // clang-format off
   //   REGION    ADDRESS                   SIZE                TYPE       WRITE   UNPRIV
-  SET_REGION( 0, FLASH_BASE_S,             SIZE_16K,           FLASH_DATA,  YES,    NO ); // Secret
-  SET_REGION( 1, FLASH_BASE_S + SIZE_16K,  SIZE_48K,           FLASH_CODE,   NO,    NO ); // Boardloader code
-  SET_REGION( 2, FLASH_BASE_S + SIZE_64K,  SIZE_4032K,         FLASH_DATA,  YES,    NO ); // Bootloader + Storage + Firmware
-  SET_REGION( 3, SRAM1_BASE_S,             SIZE_2496K,         SRAM,        YES,    NO ); // SRAM1/2/3/5
+  SET_REGION( 0, SECRET_START,             SIZE_16K,           FLASH_DATA,  YES,    NO ); // Secret
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/mpu.c]: done here: https://github.com/trezor/trezor-firmware/pull/3404/commits/2925e1870d11082b39550f236d1599f45232b45a

File: core/embed/trezorhal/stm32u5/mpu.c [hunk 2]
```diff
@@ -108,30 +109,51 @@ static void mpu_set_attributes() {
...
+#define L3_REST_SIZE                                                   \
+  (FLASH_SIZE - (FIRMWARE_START + BOOTLOADER_SIZE + BOARDLOADER_SIZE + \
+                 SECRET_SIZE + STORAGE_SIZE))
 
 void mpu_config_boardloader() {
   HAL_MPU_Disable();
   mpu_set_attributes();
   // clang-format off
   //   REGION    ADDRESS                   SIZE                TYPE       WRITE   UNPRIV
-  SET_REGION( 0, FLASH_BASE_S,             SIZE_16K,           FLASH_DATA,  YES,    NO ); // Secret
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/mpu.c:130]: done here: https://github.com/trezor/trezor-firmware/pull/3404/commits/2925e1870d11082b39550f236d1599f45232b45a

Note: successive PR https://github.com/trezor/trezor-firmware/pull/3406 adds the "initial" mpu settings that needs this treatment too. Once this is approved, i will rebase that and do the changes there.

File: core/embed/trezorhal/stm32u5/mpu.c [hunk 3]
```diff
@@ -161,13 +187,17 @@ void mpu_config_firmware() {
...
   // clang-format off
   //   REGION    ADDRESS                   SIZE                TYPE       WRITE   UNPRIV
-  SET_REGION( 0, FLASH_BASE_S + SIZE_192K, SIZE_128K,          FLASH_DATA,  YES,   YES ); // Storage
-  SET_REGION( 1, FLASH_BASE_S + SIZE_320K, SIZE_3776K,         FLASH_CODE,   NO,   YES ); // Firmware
-  SET_REGION( 2, SRAM1_BASE_S,             SIZE_2496K,         SRAM,        YES,   YES ); // SRAM1/2/3/5
-  SET_REGION( 3, GFXMMU_BUFFERS_S,         SIZE_16M,           SRAM,        YES,   YES ); // Frame buffer
-  SET_REGION( 4, PERIPH_BASE_S,            SIZE_256M,          PERIPHERAL,  YES,   YES ); // Peripherals
-  SET_REGION( 5, FLASH_OTP_BASE,           FLASH_OTP_SIZE,     FLASH_DATA,  YES,   YES ); // OTP
-  DIS_REGION( 6 );
+  SET_REGION( 0, STORAGE_START,            L3_PREV_SIZE,       FLASH_DATA,  YES,   YES ); // Storage
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/mpu.c]: done here: https://github.com/trezor/trezor-firmware/pull/3404/commits/2925e1870d11082b39550f236d1599f45232b45a

File: core/embed/models/model_D002_layout.c [hunk 4]
```diff
@@ -43,7 +43,7 @@ const flash_area_t FIRMWARE_AREA = {
     .subarea[0] =
         {
             .first_sector = 0x28,
-            .num_sectors = 456,
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/models/model_D002_layout.c:46]: done here: https://github.com/trezor/trezor-firmware/pull/3404/commits/2925e1870d11082b39550f236d1599f45232b45a

File: core/embed/trezorhal/stm32u5/mpu.c [hunk 5]
```diff
@@ -108,30 +109,51 @@ static void mpu_set_attributes() {
...
+#define L1_REST_SIZE (FLASH_SIZE - (BOARDLOADER_SIZE + SECRET_SIZE))
+
+#define L2_PREV_SIZE (SECRET_SIZE + BOARDLOADER_SIZE)
+#define L2_REST_SIZE \
+  (FLASH_SIZE - (BOOTLOADER_SIZE + BOARDLOADER_SIZE + SECRET_SIZE))
+
+#define L3_PREV_SIZE (STORAGE_SIZE)
+#define L3_REST_START (FIRMWARE_START + FIRMWARE_SIZE)
+#define L3_REST_SIZE                                                   \
+  (FLASH_SIZE - (FIRMWARE_START + BOOTLOADER_SIZE + BOARDLOADER_SIZE + \
```
[the remark this answers is not in the document]
Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/mpu.c:139]: done here: https://github.com/trezor/trezor-firmware/pull/3404/commits/2925e1870d11082b39550f236d1599f45232b45a

**Comment by TychoVrahe** [comment]
@cepetr I also added https://github.com/trezor/trezor-firmware/pull/3404/commits/568ae76b79ad72ffbc7257a246e500a42f551349 with some additional U585 modifications. Take a look.

**Comment by TychoVrahe** [comment]
added some more fixes in https://github.com/trezor/trezor-firmware/pull/3404/commits/6917c78d3adf3fa9a396c1798ea46e4a154fc884

File: core/embed/trezorhal/stm32u5/flash.c [hunk 6]
```diff
@@ -29,7 +29,13 @@
 #define FLASH_SEC_START_ADDRESS 0x0C000000
 #define FLASH_START_ADDRESS 0x08000000
 
-#define FLASH_SECTOR_COUNT (256 * 2)
+#ifdef STM32U585xx
+#define FLASH_BANK_PAGES 128
```
Comment (cepetr) [inline core/embed/trezorhal/stm32u5/flash.c:33]: We coould utilize constants from the stm32u5xxxx.h headers and not introduce our own. The code would be more universal then. See FLASH_SIZE, FLASH_BANK_SIZE, FLASH_PAGE_SIZE and FLASH_PAGE_NB.

Comment (TychoVrahe) [reply core/embed/trezorhal/stm32u5/flash.c:30]: Slightly against it, the register with flash size is somewhere behind the OTP area, which would require expanding this MPU region + adding it to boardloader.

Removed the custom FLASH_START constants, see https://github.com/trezor/trezor-firmware/commit/ecd4585ab655d0fa6cdcae6f764fa3e3739e553d

Comment (cepetr) [reply core/embed/trezorhal/stm32u5/flash.c:30]: Okay, I understand. So let the other constants be.

**Review by cepetr — APPROVED** [review]

**Pull request merged**